sdk/servicebus/service-bus/src/connectionContext.ts

```diff
@@ -180,11 +181,45 @@ export namespace ConnectionContext {
       }
     };
 
-    // Add listeners on the connection object.
-    connectionContext.connection.on(ConnectionEvents.connectionOpen, onConnectionOpen);
-    connectionContext.connection.on(ConnectionEvents.disconnected, disconnected);
-    connectionContext.connection.on(ConnectionEvents.protocolError, protocolError);
-    connectionContext.connection.on(ConnectionEvents.error, error);
+    async function refreshConnection(connectionContext: ConnectionContext) {
+      const originalConnectionId = connectionContext.connectionId;
+      try {
+        await cleanConnectionContext(connectionContext);
+      } catch {
+        log.error(
+          `[${connectionContext.connectionId}] There was an error closing the connection before reconnecting.`
+        );
+      }
+      // Create a new connection, id, locks, and cbs client.
+      connectionContext.refreshConnection();
+      addConnectionListeners(connectionContext.connection);
+      log.error(
+        `The connection "${originalConnectionId}" has been updated to "${connectionContext.connectionId}".`
+      );
+    }
+
+    function addConnectionListeners(connection: Connection) {
+      // Add listeners on the connection object.
+      connection.on(ConnectionEvents.connectionOpen, onConnectionOpen);
+      connection.on(ConnectionEvents.disconnected, disconnected);
+      connection.on(ConnectionEvents.protocolError, protocolError);
+      connection.on(ConnectionEvents.error, error);
+    }
+
+    async function cleanConnectionContext(connectionContext: ConnectionContext) {
+      // Remove listeners from the connection object.
+      connectionContext.connection.removeListener(
```

Comment:
Would it be better to remove the listeners _after_ we close the connection?  Just to make sure that if any other errors occur on the way down we still see them? Or maybe we want to stub them out perhaps with something to just capture it all locally?

Reply:
At this point we've hit a `disconnect` and have decided we don't want to use this connection anymore. I'm removing the listeners so none of them get called again while the connection is closing (I saw this sometimes trigger the disconnect event again otherwise).

If there is an error while closing the connection, `rhea-promise` will still detect that and throw it from the `connection.close()` method.

Reply:
Beautiful :) 

Yeah, I had a similar dilemma with some other code. Sounds like you're taken care of here.